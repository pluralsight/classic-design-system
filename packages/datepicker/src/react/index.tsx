export { Calendar } from './calendar'
export { CalendarDay, CalendarDayProps } from './calendar-day'
export {
  useIsInRange,
  onRangeDateSelected,
  onMultiDateSelected,
  handleDateSelectChange,
  handleRangeSelectChange
} from './utils'
export { DatePicker } from './datepicker'
