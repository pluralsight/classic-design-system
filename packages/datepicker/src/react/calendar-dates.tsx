import React, { useContext } from 'react'
import { css } from 'glamor'
import { DateObj, RenderProps } from 'dayzed'
import { HTMLPropsFor } from '@pluralsight/ps-design-system-util'
import { DateContext } from './context'
import stylesheet from '../css'

const styles = {
  calendarDate: (selected: boolean, today: boolean) =>
    css(
      stylesheet['.psds-calendar__date'],
      selected && stylesheet['.psds-calendar__date--selected'],
      today && stylesheet['.psds-calendar__date--today']
    ),
  dateFiller: () => css(stylesheet['.psds-calendar__filler'])
}

interface ChildrenRenderProps extends HTMLPropsFor<'button'> {}

interface CalendarDatesProps
  extends HTMLPropsFor<'button'>,
    Pick<RenderProps, 'getDateProps'> {
  children: (props: ChildrenRenderProps, dateObj: DateObj) => React.ReactNode
}

export const CalendarDates: React.FC<CalendarDatesProps> = ({
  getDateProps,
  children,
  ...rest
}) => {
  const calendar = useContext(DateContext)
  return (
    <>
      {calendar.weeks.map((week, weekIndex) => {
        return week.map((dateObj, index) => {
          const key = `${calendar.month}${calendar.year}${weekIndex}${index}`
          if (!dateObj) {
            return <div key={key} {...styles.dateFiller()} />
          }
          const { date, selected, selectable, today } = dateObj
          return children(
            {
              key,
              children: date.getDate(),
              disabled: !selectable,
              ...styles.calendarDate(selected, today),
              ...getDateProps({ dateObj }),
              ...rest
            },
            dateObj
          )
        })
      })}
    </>
  )
}
