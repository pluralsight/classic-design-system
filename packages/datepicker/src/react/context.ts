import { createContext } from 'react'
import { Calendar } from 'dayzed'

const initialValue = {}
// @ts-ignore: really hate this part of context and typescript
export const DateContext = createContext<Calendar>(initialValue)
