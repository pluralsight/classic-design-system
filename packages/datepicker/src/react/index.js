import { compose, css } from 'glamor'
import PropTypes from 'prop-types'
import React from 'react'

import Halo from '@pluralsight/ps-design-system-halo/react.js'
import Icon from '@pluralsight/ps-design-system-icon/react.js'
import { useTheme } from '@pluralsight/ps-design-system-theme/react.js'
import filterReactProps from '@pluralsight/ps-design-system-filter-react-props'

import stylesheet from '../css/index.js'
import {
  combineFns,
  formatDate,
  forceValidDay,
  forceValidMonth,
  forceValidYear,
  isFunction,
  parseDate
} from '../js/index.js'
import * as vars from '../vars/index.js'

import Calendar from './calendar.js'

const styles = {
  calendarContainer: _ =>
    css(stylesheet['.psds-date-picker__calendar-container']),
  datePicker: (themeName, { disabled }) =>
    compose(
      css(stylesheet['.psds-date-picker']),
      css(stylesheet[`.psds-date-picker.psds-theme--${themeName}`]),
      disabled && css(stylesheet['.psds-date-picker--disabled'])
    ),
  error: _ => compose(css(stylesheet['.psds-date-picker__error'])),
  field: _ => compose(css(stylesheet['.psds-date-picker__field'])),
  fieldContainer: (themeName, { appearance }) =>
    compose(
      css(stylesheet['.psds-date-picker__field-container']),
      css(
        stylesheet[
          `.psds-date-picker__field-container--appearance-${appearance}`
        ]
      ),
      css(
        stylesheet[
          `.psds-date-picker__field-container.psds-theme--${themeName}`
        ]
      )
    ),
  label: themeName =>
    compose(
      css(stylesheet['.psds-date-picker__label']),
      css(stylesheet[`.psds-date-picker__label.psds-theme--${themeName}`])
    ),
  icon: (themeName, { appearance }) =>
    compose(
      css(stylesheet['.psds-date-picker__icon']),
      css(stylesheet[`.psds-date-picker__icon--appearance-${appearance}`]),
      css(stylesheet[`.psds-date-picker__icon.psds-theme--${themeName}`])
    ),
  overlay: _ => css(stylesheet['.psds-date-picker__overlay']),
  subField: ({ appearance }) =>
    compose(
      css(stylesheet['.psds-date-picker__sub-field']),
      css(stylesheet[`.psds-date-picker__sub-field--appearance-${appearance}`])
    ),
  subFieldDivider: ({ appearance }) =>
    compose(
      css(stylesheet['.psds-date-picker__sub-field-divider']),
      css(
        stylesheet[
          `.psds-date-picker__sub-field-divider--appearance-${appearance}`
        ]
      )
    ),
  subLabel: themeName =>
    compose(
      css(stylesheet['.psds-date-picker__sub-label']),
      css(stylesheet[`.psds-date-picker__sub-label.psds-theme--${themeName}`])
    )
}

const DatePicker = React.forwardRef((props, ref) => {
  const themeName = useTheme()
  const value = React.useMemo(() => parseDate(props.value), [props.value])

  React.useEffect(
    function updateDateOnPropChange() {
      setDate(value)
    },
    [value]
  )

  const [date, setDate] = React.useState(value)

  const [isOpen, setIsOpen] = React.useState(false)
  const toggleIsOpen = (nextIsOpen = !isOpen) => setIsOpen(nextIsOpen)

  function handleCalendarSelect(value) {
    const nextDate = parseDate(value)
    setDate(nextDate)
    setIsOpen(false)

    if (isFunction(props.onSelect)) {
      props.onSelect(formatDate(nextDate), nextDate)
    }
  }

  function handleChange(evt) {
    const { name, value } = evt.target
    const updateRules = {
      mm: /^\d{0,2}$/,
      dd: /^\d{0,2}$/,
      yyyy: /^\d{0,4}$/
    }

    if (updateRules[name] && updateRules[name].test(value)) {
      const nextDate = { ...date, [name]: value }
      setDate(nextDate)
    }
  }

  function handleIconClick(evt) {
    evt.preventDefault()
    toggleIsOpen()
  }

  const handleKeyDown = combineFns(evt => {
    if (!isEscape(evt)) return

    evt.stopPropagation()
    evt.preventDefault()

    setIsOpen(false)
  }, props.onKeyDown)

  function handleOverlayClick(evt) {
    evt.preventDefault()
    setIsOpen(false)
  }

  function handleSubFieldBlur(evt) {
    const { name, value } = evt.target
    const forceValidValueFor = {
      dd: forceValidDay,
      mm: forceValidMonth,
      yyyy: forceValidYear
    }
    const currentDateOverwrittenByEventValue = {
      ...date,
      [name]: value
    }
    const alwaysReValidateDay = forceValidValueFor['dd'](
      currentDateOverwrittenByEventValue
    )

    const nextDate = {
      dd: alwaysReValidateDay,
      [name]: forceValidValueFor[name](currentDateOverwrittenByEventValue)
    }

    setDate(nextDate)

    if (isValidDate(nextDate) && isFunction(props.onSelect)) {
      props.onSelect(formatDate(nextDate), nextDate)
    }
  }

  function handleSubFieldFocus() {
    setIsOpen(false)
  }

  const { className, style, ...rest } = props

  return (
    <label
      {...styles.datePicker(themeName, props)}
      className={className}
      onKeyDown={handleKeyDown}
      style={style}
    >
      {props.label && (
        <div {...styles.label(themeName, props)}>{props.label}</div>
      )}

      <Halo error={props.error} gapSize={Halo.gapSizes.small}>
        <div {...styles.fieldContainer(themeName, props)}>
          <SubField
            appearance={props.appearance}
            onChange={handleChange}
            onFocus={handleSubFieldFocus}
            onBlur={handleSubFieldBlur}
            value={date.mm}
            name="mm"
            disabled={props.disabled}
            style={{ width: '32px' }}
          />

          <SubFieldDivider appearance={props.appearance} />

          <SubField
            appearance={props.appearance}
            onChange={handleChange}
            onFocus={handleSubFieldFocus}
            onBlur={handleSubFieldBlur}
            value={date.dd}
            name="dd"
            disabled={props.disabled}
            style={{ width: '28px' }}
          />

          <SubFieldDivider appearance={props.appearance} />

          <SubField
            appearance={props.appearance}
            onChange={handleChange}
            onFocus={handleSubFieldFocus}
            onBlur={handleSubFieldBlur}
            value={date.yyyy}
            name="yyyy"
            disabled={props.disabled}
            style={{ width: '50px' }}
          />

          <input
            {...styles.field(props)}
            {...filterReactProps(rest, { tagName: 'input' })}
            tabIndex="-1"
            readOnly
            disabled={props.disabled}
            ref={ref}
          />

          <button
            {...styles.icon(themeName, props)}
            disabled={props.disabled}
            onClick={handleIconClick}
          >
            <Icon id={Icon.ids.calendar} />
          </button>

          {props.error && (
            <div {...styles.error(themeName, props)}>
              <Icon id={Icon.ids.warning} />
            </div>
          )}
        </div>
      </Halo>

      {isOpen && (
        <React.Fragment>
          <Overlay onClick={handleOverlayClick} />

          <div {...styles.calendarContainer(themeName, props)}>
            <Calendar date={date} onSelect={handleCalendarSelect} />
          </div>
        </React.Fragment>
      )}

      {props.subLabel && (
        <div {...styles.subLabel(themeName, props)}>{props.subLabel}</div>
      )}
    </label>
  )
})

DatePicker.appearances = vars.appearances
export const appearances = DatePicker.appearances

DatePicker.propTypes = {
  appearance: PropTypes.oneOf(Object.values(vars.appearances)),
  className: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  label: PropTypes.node,
  onKeyDown: PropTypes.func,
  onSelect: PropTypes.func,
  style: PropTypes.object,
  subLabel: PropTypes.node,
  value: PropTypes.string
}

DatePicker.defaultProps = {
  appearance: DatePicker.appearances.default,
  disabled: false,
  error: false
}

export default DatePicker

function Overlay(props) {
  return <div {...styles.overlay()} {...props} />
}

function SubField(props) {
  const ref = React.useRef()
  const handleFocus = combineFns(() => {
    ref.current.select()
  }, props.onFocus)

  return (
    <input
      {...styles.subField(props)}
      {...filterReactProps(props, { tagName: 'input' })}
      onFocus={handleFocus}
      placeholder={props.value || props.name}
      ref={ref}
    />
  )
}

SubField.propTypes = {
  disabled: PropTypes.bool,
  name: PropTypes.string.isRequired,
  onBlur: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  onFocus: PropTypes.func.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
}

function SubFieldDivider(props) {
  return <span {...styles.subFieldDivider(props)}>/</span>
}

function isEscape(evt) {
  return evt.key === 'Escape'
}

function isValidDate({ mm, dd, yyyy }) {
  const date = new Date(yyyy, mm - 1, dd)
  const someFields = mm || dd || yyyy
  const jsDateWorks = date && date.getMonth() + 1 === mm
  return !someFields || (someFields && jsDateWorks)
}
