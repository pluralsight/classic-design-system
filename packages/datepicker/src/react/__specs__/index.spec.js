import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import * as vars from '../../vars/index.js'

import DatePicker from '../index.js'

describe('DatePicker', () => {
  describe('.appearances', () => {
    it('exists', () => expect(DatePicker.appearances).toEqual(vars.appearances))
  })

  it('renders', () => {
    const { getByTestId } = render(<DatePicker data-testid="mock-component" />)

    expect(getByTestId('mock-component')).toBeInTheDocument()
  })

  it('forwards refs', () => {
    const ref = React.createRef()

    render(<DatePicker ref={ref} />)

    expect(ref.current).not.toBeNull()
  })

  describe('onBlur prop', () => {
    it('should call onBlur when onBlur is called for a valid date', () => {
      const onBlurMock = jest.fn()
      const { getByDisplayValue } = render(
        <DatePicker value="8/14/2001" onSubBlur={onBlurMock} />
      )
      fireEvent.blur(getByDisplayValue('2001'))
      expect(onBlurMock).toHaveBeenCalledTimes(1)
      expect(onBlurMock).toHaveBeenCalledWith('8/14/2001')
    })

    it('should call onBlur when onBlur is called for an invalid date', () => {
      const onBlurMock = jest.fn()
      const { getByDisplayValue } = render(
        <DatePicker value="8/14/2001" onSubBlur={onBlurMock} />
      )
      const yearSubField = getByDisplayValue('2001')
      fireEvent.change(yearSubField, { target: { value: '' } })
      fireEvent.blur(yearSubField)
      expect(onBlurMock).toHaveBeenCalledTimes(1)
      expect(onBlurMock).toHaveBeenCalledWith(null)
    })
  })

  describe('with a controlled value', () => {
    let container
    let rerender
    const fields = { day: null, month: null, year: null }

    beforeEach(() => {
      ;({ container, rerender } = render(<DatePicker value="1/20/1993" />))

      fields.day = container.querySelector('[name="dd"]')
      fields.month = container.querySelector('[name="mm"]')
      fields.year = container.querySelector('[name="yyyy"]')
    })

    it('derives initial value from prop', () => {
      expect(fields.day.value).toEqual('20')
      expect(fields.month.value).toEqual('1')
      expect(fields.year.value).toEqual('1993')
    })

    it('should update on prop change', () => {
      rerender(<DatePicker value="8/14/2001" />)

      expect(fields.day.value).toEqual('14')
      expect(fields.month.value).toEqual('8')
      expect(fields.year.value).toEqual('2001')
    })
  })
})
