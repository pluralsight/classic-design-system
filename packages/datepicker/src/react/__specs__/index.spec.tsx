/* eslint-disable @typescript-eslint/no-unnecessary-type-assertion */
import { convertStoriesToJestCases } from '@pluralsight/ps-design-system-util'
import { render } from '@testing-library/react'
import { renderHook, act } from '@testing-library/react-hooks'
import formatISO from 'date-fns/formatISO'
import { DateObj, useDayzed } from 'dayzed'
import { axe } from 'jest-axe'
import React from 'react'

import {
  Calendar,
  CalendarDates,
  useIsInRange,
  onRangeDateSelected,
  onMultiDateSelected,
  useDateSelectChange,
  useRangeSelectChange
} from '..'
import * as stories from '../__stories__/index.story'

it('Calendar: composes className', () => {
  const Test = () => {
    const { getDateProps, ...dayzedData } = useDayzed({
      date: new Date('05/30/2020'),
      selected: new Date('05/13/2020'),
      onDateSelected: (dateObj: DateObj, evt: React.SyntheticEvent) => {}
    })

    return (
      <Calendar {...dayzedData} className="compose-classname">
        <CalendarDates getDateProps={getDateProps}>
          {renderProps => <button {...renderProps} />}
        </CalendarDates>
      </Calendar>
    )
  }

  const { container } = render(<Test />)

  expect(container.firstChild).toHaveClass('psds-calendar compose-classname')
})

it('CalendarDates: composes className', () => {
  const Test = () => {
    const { getDateProps, ...dayzedData } = useDayzed({
      date: new Date('05/30/2020'),
      selected: new Date('05/13/2020'),
      onDateSelected: (dateObj: DateObj, evt: React.SyntheticEvent) => {}
    })

    return (
      <Calendar {...dayzedData}>
        <CalendarDates
          getDateProps={getDateProps}
          className="compose-classname"
        >
          {renderProps => <button {...renderProps} />}
        </CalendarDates>
      </Calendar>
    )
  }

  const { container } = render(<Test />)

  expect(container.querySelector('.psds-calendar__filler')).toHaveClass(
    'compose-classname'
  )
  expect(container.querySelector('.psds-calendar__date')).toHaveClass(
    'compose-classname'
  )
})

test('useIsInRange: none selected', () => {
  const { result } = renderHook(() => useIsInRange([]))
  expect(result.current.isInRange(new Date('06/01/2020'))).toEqual('')
})

test('useIsInRange: start date selected', () => {
  const selected = [new Date('06/01/2020')]
  const { result } = renderHook(() => useIsInRange(selected))
  act(() => {
    result.current.onMouseEnter(new Date('06/10/2020'))
  })
  expect(result.current.isInRange(new Date('06/01/2020'))).toMatchSnapshot(
    'start date'
  )
  expect(result.current.isInRange(new Date('06/06/2020'))).toMatchSnapshot(
    'in range'
  )
  expect(result.current.isInRange(new Date('06/15/2020'))).toMatchSnapshot(
    'out of range'
  )
})

test('useIsInRange: start date and end date selected', () => {
  const selected = [new Date('06/01/2020'), new Date('06/10/2020')]
  const { result } = renderHook(() => useIsInRange(selected))
  expect(result.current.isInRange(new Date('06/01/2020'))).toMatchSnapshot(
    'start date'
  )
  expect(result.current.isInRange(new Date('06/06/2020'))).toMatchSnapshot(
    'in range'
  )
  expect(result.current.isInRange(new Date('06/15/2020'))).toMatchSnapshot(
    'out of range'
  )
})

test('onRangeDateSelected: onSelect', () => {
  const onSelect = jest.fn()
  const { result } = renderHook(() => React.useState<Date[] | undefined>())
  act(() => {
    const selectedDate = new Date('06/20/2020')
    onRangeDateSelected({
      selected: result.current[0],
      setSelected: result.current[1],
      onSelect
    })(
      {
        date: selectedDate,
        nextMonth: false,
        prevMonth: false,
        selectable: false,
        selected: true,
        today: false
      },
      {} as React.SyntheticEvent
    )
  })
  expect(onSelect).not.toHaveBeenCalled()
  const dateObj = {
    date: new Date('06/20/2020'),
    nextMonth: false,
    prevMonth: false,
    selectable: true,
    selected: true,
    today: false
  }
  act(() => {
    onRangeDateSelected({
      selected: result.current[0],
      setSelected: result.current[1],
      onSelect
    })(dateObj, {} as React.SyntheticEvent)
  })
  expect(onSelect).toHaveBeenCalledWith({}, dateObj)
})

test('onRangeDateSelected: selectable true', () => {
  const { result } = renderHook(() => React.useState<Date[] | undefined>())
  const selectedDate = new Date('06/20/2020')
  const dateObj = {
    date: selectedDate,
    nextMonth: false,
    prevMonth: false,
    selectable: true,
    selected: true,
    today: false
  }
  act(() => {
    onRangeDateSelected({
      selected: result.current[0],
      setSelected: result.current[1]
    })(dateObj, {} as React.SyntheticEvent)
  })
  expect(result.current[0]).toStrictEqual([selectedDate])
})

test('onRangeDateSelected: start date selects end date', () => {
  const startDate = new Date('06/10/2020')
  const { result } = renderHook(() =>
    React.useState<Date[] | undefined>([startDate])
  )
  const selectedDate = new Date('06/20/2020')
  const dateObj = {
    date: selectedDate,
    nextMonth: false,
    prevMonth: false,
    selectable: true,
    selected: true,
    today: false
  }
  act(() => {
    onRangeDateSelected({
      selected: result.current[0],
      setSelected: result.current[1]
    })(dateObj, {} as React.SyntheticEvent)
  })
  expect(result.current[0]).toStrictEqual([startDate, selectedDate])
})

test('onRangeDateSelected: end date selects start date', () => {
  const initialDates = [new Date('06/10/2020'), new Date('06/20/2020')]
  const { result } = renderHook(() =>
    React.useState<Date[] | undefined>(initialDates)
  )
  const selectedDate = new Date('06/30/2020')
  const dateObj = {
    date: selectedDate,
    nextMonth: false,
    prevMonth: false,
    selectable: true,
    selected: true,
    today: false
  }
  act(() => {
    onRangeDateSelected({
      selected: result.current[0],
      setSelected: result.current[1]
    })(dateObj, {} as React.SyntheticEvent)
  })
  expect(result.current[0]).toStrictEqual([selectedDate])
})

test('onMultiDateSelected: onSelect', () => {
  const onSelect = jest.fn()
  const { result } = renderHook(() => React.useState<Date[] | undefined>())
  act(() => {
    const selectedDate = new Date('06/20/2020')
    onMultiDateSelected({
      selected: result.current[0],
      setSelected: result.current[1],
      onSelect
    })(
      {
        date: selectedDate,
        nextMonth: false,
        prevMonth: false,
        selectable: false,
        selected: true,
        today: false
      },
      {} as React.SyntheticEvent
    )
  })
  expect(onSelect).not.toHaveBeenCalled()
  const dateObj = {
    date: new Date('06/20/2020'),
    nextMonth: false,
    prevMonth: false,
    selectable: true,
    selected: true,
    today: false
  }
  act(() => {
    onMultiDateSelected({
      selected: result.current[0],
      setSelected: result.current[1],
      onSelect
    })(dateObj, {} as React.SyntheticEvent)
  })
  expect(onSelect).toHaveBeenCalledWith({}, dateObj)
})

test('onMultiDateSelected: selectable (toggle test)', () => {
  const selectedDate = new Date('06/20/2020')
  const { result } = renderHook(() => React.useState<Date[] | undefined>())
  const dateObj = {
    date: selectedDate,
    nextMonth: false,
    prevMonth: false,
    selectable: true,
    selected: false,
    today: false
  }
  act(() => {
    onMultiDateSelected({
      selected: result.current[0],
      setSelected: result.current[1]
    })(dateObj, {} as React.SyntheticEvent)
  })
  expect(result.current[0]).toStrictEqual([selectedDate])
  act(() => {
    onMultiDateSelected({
      selected: result.current[0],
      setSelected: result.current[1]
    })({ ...dateObj, selected: true }, {} as React.SyntheticEvent)
  })
  expect(result.current[0]).toStrictEqual([])
})

test('useDateSelectChange', () => {
  const { result } = renderHook(() => {
    const [selected, setSelected] = React.useState<Date | undefined>()
    const [slide, setSlide] = React.useState<
      'forward' | 'backward' | undefined
    >()
    const [value, onChange] = useDateSelectChange({
      setSelected,
      selected,
      setSlide
    })
    return { selected, slide, value, onChange }
  })
  expect(result.current.value).toBe('')
  expect(result.current.selected).toBe(undefined)
  expect(result.current.slide).toBe(undefined)
  act(() => {
    result.current.onChange({
      target: { value: '1/10' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.value).toBe('1/10')
  expect(result.current.selected).toBe(undefined)
  expect(result.current.slide).toBe(undefined)
  act(() => {
    result.current.onChange({
      target: { value: '01/10/2020' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.value).toBe('01/10/2020')
  expect(formatISO(result.current.selected!)).toBe(
    formatISO(new Date('01/10/2020'))
  )
  expect(result.current.slide).toBe(undefined)
  act(() => {
    result.current.onChange({
      target: { value: '08/10/2020' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.value).toBe('08/10/2020')
  expect(formatISO(result.current.selected!)).toBe(
    formatISO(new Date('08/10/2020'))
  )
  expect(result.current.slide).toBe('forward')
  act(() => {
    result.current.onChange({
      target: { value: '03/10/2020' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.value).toBe('03/10/2020')
  expect(formatISO(result.current.selected!)).toBe(
    formatISO(new Date('03/10/2020'))
  )
  expect(result.current.slide).toBe('backward')
  act(() => {
    result.current.onChange({
      target: { value: '13/10/2020' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.value).toBe('13/10/2020')
  expect(formatISO(result.current.selected!)).toBe(
    formatISO(new Date('03/10/2020'))
  )
  expect(result.current.slide).toBeUndefined()
  act(() => {
    result.current.onChange({
      target: { value: '03/11/2020' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.value).toBe('03/11/2020')
  expect(formatISO(result.current.selected!)).toBe(
    formatISO(new Date('03/11/2020'))
  )
  expect(result.current.slide).toBe(undefined)
})

test('useDateSelectChange: custom date format', () => {
  const { result } = renderHook(() => {
    const [selected, setSelected] = React.useState<Date | undefined>()
    const [slide, setSlide] = React.useState<
      'forward' | 'backward' | undefined
    >()
    const dateFormat = 'yyyy-MM-dd'
    const [value, onChange] = useDateSelectChange({
      setSelected,
      selected,
      setSlide,
      dateFormat
    })
    return { selected, slide, value, onChange }
  })
  act(() => {
    result.current.onChange({
      target: { value: '01/10/2020' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.value).toBe('01/10/2020')
  expect(result.current.selected).toBe(undefined)
  act(() => {
    result.current.onChange({
      target: { value: '2020-01-10' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.value).toBe('2020-01-10')
  expect(formatISO(result.current.selected!)).toBe(
    formatISO(new Date('01/10/2020'))
  )
})

test('useRangeSelectChange', () => {
  const { result } = renderHook(() => {
    const [selected, setSelected] = React.useState<Date[] | undefined>()
    const [slide, setSlide] = React.useState<
      'forward' | 'backward' | undefined
    >()
    const [startValue, onStartChange] = useRangeSelectChange({
      start: true,
      setSelected,
      selected,
      setSlide
    })
    const [endValue, onEndChange] = useRangeSelectChange({
      start: false,
      setSelected,
      selected,
      setSlide
    })
    return {
      selected,
      slide,
      startValue,
      onStartChange,
      endValue,
      onEndChange
    }
  })
  expect(result.current.startValue).toBe('')
  expect(result.current.endValue).toBe('')
  expect(result.current.selected).toBe(undefined)
  expect(result.current.slide).toBe(undefined)
  act(() => {
    result.current.onEndChange({
      target: { value: '03/11' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.startValue).toBe('')
  expect(result.current.endValue).toBe('03/11')
  expect(result.current.selected).toBe(undefined)
  expect(result.current.slide).toBe(undefined)
  act(() => {
    result.current.onStartChange({
      target: { value: '03/01' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.startValue).toBe('03/01')
  expect(result.current.endValue).toBe('03/11')
  expect(result.current.selected).toBe(undefined)
  expect(result.current.slide).toBe(undefined)
  act(() => {
    result.current.onStartChange({
      target: { value: '03/01' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  act(() => {
    result.current.onEndChange({
      target: { value: '03/11/2021' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.startValue).toBe('03/01')
  expect(result.current.endValue).toBe('03/11/2021')
  expect(result.current.selected).toStrictEqual([])
  expect(result.current.slide).toBe(undefined)
  act(() => {
    result.current.onStartChange({
      target: { value: '03/01/2020' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  act(() => {
    result.current.onEndChange({
      target: { value: '03/11' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.selected!.map(date => formatISO(date))).toMatchObject(
    [new Date('03/01/2020')].map(date => formatISO(date))
  )
  expect(result.current.startValue).toBe('03/01/2020')
  expect(result.current.endValue).toBe('03/11')
  expect(result.current.slide).toBe(undefined)
  act(() => {
    result.current.onEndChange({
      target: { value: '03/11/2020' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.selected!.map(date => formatISO(date))).toMatchObject(
    [new Date('03/01/2020'), new Date('03/11/2020')].map(date =>
      formatISO(date)
    )
  )
  expect(result.current.startValue).toBe('03/01/2020')
  expect(result.current.endValue).toBe('03/11/2020')
  expect(result.current.slide).toBe(undefined)
  act(() => {
    result.current.onStartChange({
      target: { value: '03/05/2020' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  act(() => {
    result.current.onEndChange({
      target: { value: '03/21/2020' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.selected!.map(date => formatISO(date))).toMatchObject(
    [new Date('03/05/2020'), new Date('03/21/2020')].map(date =>
      formatISO(date)
    )
  )
  expect(result.current.startValue).toBe('03/05/2020')
  expect(result.current.endValue).toBe('03/21/2020')
  expect(result.current.slide).toBe(undefined)
  act(() => {
    result.current.onEndChange({
      target: { value: '05/21/2020' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.selected!.map(date => formatISO(date))).toMatchObject(
    [new Date('03/05/2020'), new Date('05/21/2020')].map(date =>
      formatISO(date)
    )
  )
  expect(result.current.startValue).toBe('03/05/2020')
  expect(result.current.endValue).toBe('05/21/2020')
  expect(result.current.slide).toBe('forward')
  act(() => {
    result.current.onStartChange({
      target: { value: '02/05/2020' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.selected!.map(date => formatISO(date))).toMatchObject(
    [new Date('02/05/2020'), new Date('05/21/2020')].map(date =>
      formatISO(date)
    )
  )
  expect(result.current.startValue).toBe('02/05/2020')
  expect(result.current.endValue).toBe('05/21/2020')
  expect(result.current.slide).toBe('backward')
})

describe('accessibility', () => {
  const cases = convertStoriesToJestCases(stories)
  describe.each(cases)('%s story', (_name, Story) => {
    it('has no axe-core violations', async () => {
      const { container } = render(<Story {...Story.args} />)
      const results = await axe(container)
      expect(results).toHaveNoViolations()
    })
  })
})

test('useRangeSelectChange: custom date format', () => {
  const { result } = renderHook(() => {
    const [selected, setSelected] = React.useState<Date[] | undefined>()
    const [slide, setSlide] = React.useState<
      'forward' | 'backward' | undefined
    >()
    const dateFormat = 'yyyy-MM-dd'
    const [startValue, onStartChange] = useRangeSelectChange({
      start: true,
      setSelected,
      selected,
      setSlide,
      dateFormat
    })
    const [endValue, onEndChange] = useRangeSelectChange({
      start: false,
      setSelected,
      selected,
      setSlide,
      dateFormat
    })
    return {
      selected,
      slide,
      startValue,
      onStartChange,
      endValue,
      onEndChange
    }
  })
  act(() => {
    result.current.onStartChange({
      target: { value: '03/01/2021' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  act(() => {
    result.current.onEndChange({
      target: { value: '03/11/2021' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.startValue).toBe('03/01/2021')
  expect(result.current.endValue).toBe('03/11/2021')
  expect(result.current.selected).toStrictEqual(undefined)
  expect(result.current.slide).toBe(undefined)
  act(() => {
    result.current.onStartChange({
      target: { value: '2021-03-01' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  act(() => {
    result.current.onEndChange({
      target: { value: '2021-03-11' }
    } as React.ChangeEvent<HTMLInputElement>)
  })
  expect(result.current.startValue).toBe('2021-03-01')
  expect(result.current.endValue).toBe('2021-03-11')
  expect(result.current.selected!.map(date => formatISO(date))).toMatchObject([
    formatISO(new Date('03/01/2021')),
    formatISO(new Date('03/11/2021'))
  ])
})
