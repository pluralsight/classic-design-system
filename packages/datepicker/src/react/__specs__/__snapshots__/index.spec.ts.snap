// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useIsInRange: start date and end date selected: in range 1`] = `
Object {
  "data-css-4a5f34": "",
}
`;

exports[`useIsInRange: start date and end date selected: out of range 1`] = `Object {}`;

exports[`useIsInRange: start date and end date selected: start date 1`] = `
Object {
  "data-css-1r79e04": "",
}
`;

exports[`useIsInRange: start date selected: in range 1`] = `
Object {
  "data-css-4a5f34": "",
}
`;

exports[`useIsInRange: start date selected: out of range 1`] = `Object {}`;

exports[`useIsInRange: start date selected: start date 1`] = `
Object {
  "data-css-1r79e04": "",
}
`;
